src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,363 @@
+
+
+'''
+>>> issue_cert
+{'name': 'check_cert', 'tos': True, 'csr_id': 14, 'acme_directory_uri': 'https://acme-staging-v02.api.letsencrypt.org/directory', 'create_type': 'CERTIFICATE_CREATE_ACME', 'dns_mapping': {'acmedev.agencialivre.com.br': 1}}
+'''
+
+'''
+TODO'S:
+1) IS THERE A MIN KEY SIZE REQUIRED BY LETS ENCRYPT IN CSR - HANDLE THE EXCEPTION GRACEFULLY
+2) CHECK IF _GET_INSTANCE CAN BE CALLED FROM MIDDLEWARE.CALL_SYNC
+3) Domain names should not end in periods ? research
+4) Integrate alerts
+5) See what can be done to respect rate limits
+'''
+
+
+class ACMERegistrationService(CRUDService):
+
+    class Config:
+        datastore = 'system.acmeregistration'
+        datastore_extend = 'acme.registration.register_extend'
+        namespace = 'acme.registration'
+        private = True
+
+    async def register_extend(self, data):
+        data['body'] = {
+            key: value for key, value in
+            (await self.middleware.call(
+                'datastore.query', 'system.acmeregistrationbody',
+                [['acme', '=', data['id']]], {'get': True}
+            )).items() if key != 'acme'
+        }
+        return data
+
+    def get_directory(self, acme_directory_uri):
+        try:
+            response = requests.get(acme_directory_uri).json()
+            return messages.Directory({
+                key: response[key] for key in ['newAccount', 'newNonce', 'newOrder', 'revokeCert']
+            })
+        except (requests.ConnectionError, requests.Timeout, json.JSONDecodeError, KeyError) as e:
+            raise CallError(f'Unable to retrieve directory : {e}')
+
+    @accepts(
+        Dict(
+            'acme_registration_create',
+            Bool('tos', default=False),
+            Dict(
+                'JWK_create',
+                Int('key_size', default=2048),
+                Int('public_exponent', default=65537)
+            ),
+            Str('acme_directory_uri', required=True),
+        )
+    )
+    def do_create(self, data):
+
+        # STEPS FOR CREATION
+        # 1) CREATE KEY
+        # 2) REGISTER CLIENT
+        # 3) SAVE REGISTRATION OBJECT
+        # 4) SAVE REGISTRATION BODY
+
+        verrors = ValidationErrors()
+
+        directory = self.get_directory(data['acme_directory_uri'])
+        if not isinstance(directory, messages.Directory):
+            verrors.add(
+                'acme_registration_create.direcotry_uri',
+                f'System was unable to retrieve the directory with the specified acme_directory_uri: {directory}'
+            )
+
+        if not data['tos']:
+            verrors.add(
+                'acme_registration_create.tos',
+                'Please agree to the terms of service'
+            )
+
+        # For now we assume that only root is responsible for certs issued under ACME protocol
+        email = (self.middleware.call_sync('user.query', [['id', '=', 1]]))[0]['email']
+        if not email:
+            verrors.add(
```

Comment:
I think this should be a `CallError` because it is not related to the passed arguments